### ISSUE TITLE
git_revwalk outputs unwanted commits

### ISSUE THREAD

**Issue by ltrzesniewski** [libgit2/libgit2#4333]
Under some circumstances, `git_revwalk` will output too many commits. The behavior depends on the order of `git_revwalk_hide` calls.

I originally spotted the problem with a program which does many `git_revwalk_push` followed by many `git_revwalk_hide` in a single revwalk (in order to find out which commits are new after a fetch on every branch), but I managed to extract a test case which shows the issue. I'm not sure what _exactly_ is the root cause though.

Here's what the [test repo](https://github.com/libgit2/libgit2/files/1263157/test.git.zip) looks like:

![image](https://user-images.githubusercontent.com/7913492/29867585-56b9705a-8d7c-11e7-95ae-7980db584d44.png)

The test program does the equivalent of the following commands:

```
$ git rev-list D ^A ^B ^E
790ba0facf6fd103699a5c40cd19dad277ff49cd
b82cee5004151ae0c4f82b69fb71b87477664b6f

$ git rev-list D ^B ^A ^E
790ba0facf6fd103699a5c40cd19dad277ff49cd
b82cee5004151ae0c4f82b69fb71b87477664b6f
```

Which produce the same correct result with git, but libgit2 outputs more commits in the second case.

### Reproduction steps

Add the following file to the examples directory (for instance as `bug-repro.c`):

```C
#include "common.h"

static void dump_walk(git_revwalk *walk)
{
	git_oid oid;
	char buf[GIT_OID_HEXSZ+1];

	while (!git_revwalk_next(&oid, walk)) {
		git_oid_fmt(buf, &oid);
		buf[GIT_OID_HEXSZ] = '\0';
		printf("%s\n", buf);
	}
}

int main(void)
{
	git_repository *repo;
	git_revwalk *walk;

	git_libgit2_init();

	check_lg2(git_repository_open_ext(&repo, "test.git", 0, NULL), "opening repository", NULL);
	check_lg2(git_revwalk_new(&walk, repo), "allocating revwalk", NULL);

	puts("Normal behavior:");

	git_revwalk_push_ref(walk, "refs/heads/D");
	git_revwalk_hide_ref(walk, "refs/heads/A");
	git_revwalk_hide_ref(walk, "refs/heads/B");
	git_revwalk_hide_ref(walk, "refs/heads/E");

	dump_walk(walk);

	git_revwalk_free(walk);
	check_lg2(git_revwalk_new(&walk, repo), "reallocating revwalk", NULL);

	puts("Buggy behavior:");
	
	git_revwalk_push_ref(walk, "refs/heads/D");
	git_revwalk_hide_ref(walk, "refs/heads/B");
	git_revwalk_hide_ref(walk, "refs/heads/A");
	git_revwalk_hide_ref(walk, "refs/heads/E");

	dump_walk(walk);

	git_libgit2_shutdown();
	return 0;
}
```

Download and extract [test.git.zip](https://github.com/libgit2/libgit2/files/1263157/test.git.zip) to the program's build output directory, then execute the program.

### Expected behavior

```
Normal behavior:
790ba0facf6fd103699a5c40cd19dad277ff49cd
b82cee5004151ae0c4f82b69fb71b87477664b6f
Buggy behavior:
790ba0facf6fd103699a5c40cd19dad277ff49cd
b82cee5004151ae0c4f82b69fb71b87477664b6f
```

### Actual behavior

```
Normal behavior:
790ba0facf6fd103699a5c40cd19dad277ff49cd
b82cee5004151ae0c4f82b69fb71b87477664b6f
Buggy behavior:
790ba0facf6fd103699a5c40cd19dad277ff49cd
b82cee5004151ae0c4f82b69fb71b87477664b6f
743398b425d6c216d6cfaae3786b5bc436393ae5
9241da0deabd1c7fe9ceb1691c77186abdbf500b
8aa93677ae58738ff998c9cced39e988e1689792
```

Note that if you replace `git_revwalk_free`/`git_revwalk_new` with `git_revwalk_reset`, the problem goes away.

### Version of libgit2 (release number or SHA1)

Current master: 3c21645

### Operating system(s) tested

- Originally encountered with LibGit2Sharp 0.24.0 under Windows 10
- Reproduced under Windows Subsystem for Linux (Windows 10)

**Comment by ltrzesniewski** [comment]
FWIW this bug is still present as of 0eca423 (>v0.27.0)

**Comment by pks-t** [comment]
Thanks for this great issue report, and sorry for not tackling it earlier. I can reproduce the error easily and am investigating the cause

**Comment by pks-t** [comment]
Could you please give #4622 a go?

**Comment by ltrzesniewski** [comment]
Thanks a lot!

This fixes the issue in the example I provided. I ran a couple additional tests by changing the refs in the example repo and everything worked fine.

I'd need to wait for this change to end up in libgit2sharp in order to test it in my real app though. I can either wait if you think you can merge this without risk, or build a custom libgit2sharp version to test this. In either way, I'd need to let the app run for a few days on our quite busy repository in order to see if something weird shows up.

**Comment by ethomson** [comment]
Thanks for the wonderful bug report.

**Comment by ethomson** [comment]
Closed via #4622
